apt/method.go

```diff
@@ -0,0 +1,225 @@
+package apt
+
+import (
+	"bufio"
+	"context"
+	"crypto/md5"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+)
+
+func NewAptMethod(output io.Writer, input *bufio.Reader) *AptMethod {
+	return &AptMethod{
+		config: &AptMethodConfig{},
+		writer: NewAptMessageWriter(output),
+		reader: NewAptMessageReader(input),
+	}
+}
+
+// httpClient exists to enable mocking of http.Client.
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// downloader exists to enable mocking of AptMethod.download.
+type downloader interface {
+	download(io.ReadCloser, string) (string, error)
+}
+
+type realDownloader struct{}
+
+type AptMethod struct {
+	config *AptMethodConfig
+	writer *AptMessageWriter
+	reader *AptMessageReader
+	client httpClient
+	dl     downloader
+}
+
+type AptMethodConfig struct {
+	serviceAccountJSON, serviceAccountEmail string
+}
+
+func (m *AptMethod) Run(ctx context.Context) {
+	m.writer.SendCapabilities()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+		msg, err := m.reader.ReadMessage(ctx)
+		if err != nil {
+			// TODO: log?
+			continue
+		}
+		switch msg.code {
+		case 600:
+			m.handleAcquire(msg)
+		case 601:
+			m.handleConfigure(msg)
+		default:
+			m.writer.Fail("Unsupported API")
+		}
+	}
+}
+
+func (m *AptMethod) getClient() error {
+	if m.client != nil {
+		return nil
+	}
+
+	var ts oauth2.TokenSource
+	ctx := context.Background()
+	switch {
+	case m.config.serviceAccountJSON != "":
+		json, err := ioutil.ReadFile(m.config.serviceAccountJSON)
+		if err == nil {
+			creds, ierr := google.CredentialsFromJSON(ctx, json)
+			if ierr == nil {
+				ts = creds.TokenSource
+			}
+		}
+	case m.config.serviceAccountEmail != "":
+		ts = google.ComputeTokenSource(m.config.serviceAccountEmail)
+	default:
+		creds, err := google.FindDefaultCredentials(ctx)
+		if err == nil {
+			ts = creds.TokenSource
+		}
+	}
+	if ts == nil {
+		return errors.New("Failed to obtain token source")
+	}
+	m.client = oauth2.NewClient(ctx, ts)
+	return nil
+}
+
+func (r realDownloader) download(body io.ReadCloser, filename string) (string, error) {
```

Comment:
it's not clear what the returned string represents without reading func implementation. Add func docs describing what is returned